### PR TITLE
updated WrapEllipsoid

### PR DESCRIPTION
**Pull request by pariterre**
### Brief summary of changes
Updated WrapEllipsoid implementation according to new OpenSim 4.0 API

### Testing I've completed
They should all pass

### Looking for feedback on...
This PR should mimic the previous one on Cylinders

### CHANGELOG.md (choose one)
- no need to update because CHANGELOG.md's text already fits these changes (after PR on Cylinders is accepted)

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2299)

### CODE REVIEW

File: OpenSim/Simulation/Wrap/WrapEllipsoid.h [hunk 1]
```diff
@@ -51,50 +42,40 @@ class WrapResult;
...
+public:
 //=============================================================================
-// DATA
+// PROPERTIES
 //=============================================================================
-protected:
-
-    PropertyDblArray _dimensionsProp;
-    Array<double>& _dimensions;
+    OpenSim_DECLARE_PROPERTY(dimensions, SimTK::Vec3, "The length of the radii of the ellipsoid.");
```
Comment (chrisdembia) [inline OpenSim/Simulation/Wrap/WrapEllipsoid.h]: Please limit lines to 80 characters.

File: OpenSim/Simulation/Wrap/WrapEllipsoid.cpp [hunk 2]
```diff
@@ -276,14 +222,14 @@ int WrapEllipsoid::wrapLine(const SimTK::State& s, SimTK::Vec3& aPoint1, SimTK::
     // the ellipsoid dimensions because they do not change from one call to the
     // next. You don't want the factor to change because the algorithm uses
     // some vectors (r1, r2, c1) from the previous call.
-    aWrapResult.factor = 3.0 / (_dimensions[0] + _dimensions[1] + _dimensions[2]);
+    aWrapResult.factor = 3.0 / (get_dimensions()[0] + get_dimensions()[1] + get_dimensions()[2]);
```
Comment (chrisdembia) [inline OpenSim/Simulation/Wrap/WrapEllipsoid.cpp]: You could use `get_dimensions().sum()`.

File: OpenSim/Simulation/Wrap/WrapEllipsoid.cpp [hunk 3]
```diff
@@ -158,19 +130,10 @@ void WrapEllipsoid::extendScale(const SimTK::State& s, const ScaleSet& scaleSet)
...
-//_____________________________________________________________________________
-/**
-* Copy data members from one WrapEllipsoid to another.
-*
-* @param aWrapEllipsoid WrapEllipsoid to be copied.
-*/
-void WrapEllipsoid::copyData(const WrapEllipsoid& aWrapEllipsoid)
-{
-    _dimensions = aWrapEllipsoid._dimensions;
+        previousDimensions[i] *= localScaleVector[i].norm();
```
Comment (chrisdembia) [inline OpenSim/Simulation/Wrap/WrapEllipsoid.cpp]: You can also use `upd_dimensions()[i]`.

File: OpenSim/Simulation/Wrap/WrapEllipsoid.cpp [hunk 4]
```diff
@@ -72,62 +69,37 @@ WrapEllipsoid::~WrapEllipsoid()
...
     // Base class
-    WrapObject::connectToModelAndBody(aModel, aBody);
+    WrapObject::extendFinalizeFromProperties();
 
     // maybe set a parent pointer, _body = aBody;
 
-    if (_dimensions[0] < 0.0 || _dimensions[1] < 0.0 || _dimensions[2] < 0.0)
+    if (get_dimensions()[0] < 0.0 || get_dimensions()[1] < 0.0 || get_dimensions()[2] < 0.0)
     {
         string errorMessage = "Error: dimensions for WrapEllipsoid " + getName() + " were either not specified, or are negative.";
```
Comment (chrisdembia) [inline OpenSim/Simulation/Wrap/WrapEllipsoid.cpp]: This error message is no longer correct: unspecified dimensions are positive and do not cause this conditional to be true.

File: OpenSim/Simulation/Wrap/WrapEllipsoid.cpp [hunk 5]
```diff
@@ -72,62 +69,37 @@ WrapEllipsoid::~WrapEllipsoid()
...
+void WrapEllipsoid::constructProperties()
 {
     // BASE CLASS
     //WrapObject::setupProperties();
 
-    const double defaultDimensions[] = {-1.0, -1.0, -1.0};
-    _dimensionsProp.setName("dimensions");
-    _dimensionsProp.setValue(3, defaultDimensions);
-    _propertySet.append(&_dimensionsProp);
+    SimTK::Vec3 defaultDimensions = {1.0, 1.0, 1.0};
```
Comment (chrisdembia) [inline OpenSim/Simulation/Wrap/WrapEllipsoid.cpp]: These dimensions of 1.0 meter are huge. Consider using defaults of 1 cm (0.01).

File: OpenSim/Simulation/Wrap/WrapEllipsoid.cpp (hunk 2 above)
Comment (pariterre) [reply OpenSim/Simulation/Wrap/WrapEllipsoid.cpp]: Thank for the tip :)

File: OpenSim/Simulation/Wrap/WrapEllipsoid.cpp (hunk 5 above)
Comment (pariterre) [reply OpenSim/Simulation/Wrap/WrapEllipsoid.cpp]: 1 cm seems a bit tiny, I made them 5 cm to match the outer radius of the torus

**Comment by jimmyDunne** [comment]
@pariterre, can you do update with master and test? If it passes we can merge.

**Comment by pariterre** [comment]
@jimmyDunne seems good :)

**Comment by jimmyDunne** [comment]
I pulled this remote branch, built, and tested in Matlab. Everything past locally and ran in Matlab without issue. 

@ayman asked if the XML file would be changed at all. The difference is an additional comment for the dimensions, which doesn't warrant an XML version number update. 

Thanks, @pariterre!

**Pull request merged**